### PR TITLE
enable operator gpu unittest

### PR DESCRIPTION
**Pull request by QiJune**

### CODE REVIEW

File: paddle/operators/add_op.cu [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#define EIGEN_USE_GPU
```
Comment (QiJune) [inline paddle/operators/add_op.cu]: Have to add this macro, otherwise, will cause segment default

File: paddle/pybind/tensor_bind.h [hunk 2]
```diff
@@ -77,19 +81,39 @@ inline py::buffer_info CastToPyBuffer(framework::Tensor &tensor) {
...
 
   self.Resize(framework::make_ddim(dims));
-  auto *dst = self.mutable_data<T>(paddle::platform::CPUPlace());
+  auto *dst = self.mutable_data<T>(place);
   std::memcpy(dst, array.data(), sizeof(T) * array.size());
 }
 
+#ifndef PADDLE_ONLY_CPU
+template <typename T>
+void PyCUDATensorSetFromArray(
```
Comment (QiJune) [inline paddle/pybind/tensor_bind.h]: partial specialization of function template is not allowed. So just define another function here

File: python/paddle/v2/framework/tests/op_test_util.py [hunk 3]
```diff
@@ -25,41 +25,48 @@ def test_all(self):
...
-                    var = scope.create_var(in_name).get_tensor()
-                    arr = getattr(self, in_name)
-                    var.set_dims(arr.shape)
-                    var.set(arr)
-                else:
-                    kwargs[in_name] = "@EMPTY@"
+            places = []
+            places.append(core.CPUPlace())
+            if core.is_compile_gpu():
+                places.append(core.GPUPlace(0))
```
Comment (QiJune) [inline python/paddle/v2/framework/tests/op_test_util.py]: Run CPU first, and then GPU

File: python/paddle/v2/framework/tests/test_add_two_op.py [hunk 4]
```diff
@@ -8,8 +8,8 @@ class TestAddOp(unittest.TestCase):
 
     def setUp(self):
         self.type = "add_two"
-        self.X = numpy.random.random((342, 345)).astype("float32")
-        self.Y = numpy.random.random((342, 345)).astype("float32")
+        self.X = numpy.random.random((102, 105)).astype("float32")
```
Comment (QiJune) [inline python/paddle/v2/framework/tests/test_add_two_op.py]: Sometimes, GPU memory is not enough for unit-test, so reduce the size

File: cmake/flags.cmake [hunk 5]
```diff
@@ -9,6 +9,11 @@ function(CheckCompilerCXX11Flag)
         if(${CMAKE_CXX_COMPILER_VERSION} VERSION_LESS 4.8)
             message(FATAL_ERROR "Unsupported GCC version. GCC >= 4.8 required.")
         endif()
+        # TODO(qijun) gcc 4.9 or later versions raise SEGV due to the optimization problem.
+        # Use Debug mode instead for now.
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 4.9 OR CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL 4.9) 
+            set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "" FORCE)
```
Comment (QiJune) [inline cmake/flags.cmake]: If gcc 4.8 and nvcc 8.0, both debug and release will be fine.
If gcc 5.4 and nvcc 8.0, debug is fine, but release will cause segment fault.

File: paddle/framework/detail/tensor-inl.h [hunk 6]
```diff
@@ -62,9 +61,11 @@ inline T* Tensor::mutable_data(platform::Place place) {
     if (platform::is_cpu_place(place)) {
       holder_.reset(new PlaceholderImpl<T, platform::CPUPlace>(
           boost::get<platform::CPUPlace>(place), size));
+    } else if (platform::is_gpu_place(place)) {
+#ifdef PADDLE_ONLY_CPU
+      PADDLE_THROW("'GPUPlace' is not supported in CPU only device.");
     }
```
Comment (gangliao) [inline paddle/framework/detail/tensor-inl.h]: remove }

File: cmake/flags.cmake (hunk 5 above)
Comment (wangkuiyi) [inline cmake/flags.cmake]: How about we constraint GCC version in Dockerfile, other than checking here? 

I know checking in CMake provides a guarantee, but I am afraid adding too many such constraints would complicate our building system too much.

Comment (QiJune) [inline cmake/flags.cmake]: Yes, I will  constraint GCC version in Dockerfile.
But some others may compile paddle in gcc5.4 environment, so, the unittest test_framework will fall.

File: paddle/framework/detail/tensor-inl.h (hunk 6 above)
Comment (QiJune) [inline paddle/framework/detail/tensor-inl.h]: { is outside of macro, so } is needed inside macro

**Pull request merged**